src/operator/warp.cu

```diff
@@ -0,0 +1,433 @@
+/*!
+ * Copyright [2016] <Contributors>
+ * \file warp.cu
+ * \brief warp operator
+ * \author Xu Dong
+*/
+#include "./warp-inl.h"
+#include <mshadow/tensor.h>
+#include <mshadow/cuda/reduce.cuh>
+#include <algorithm>
+#include <vector>
+
+/* Warp is done in BHWD (coalescing is not obvious in BDHW)
+   we assume BHWD format in inputImages
+   we assume BHW(YX) format on grids
+*/
+namespace mshadow {
+namespace warp_cuda {
+
+__device__ void getTopLeft(double x, int xOut, int width, int& point, double& weight) {
+  double xcoord = x + xOut;
+  point = floor(xcoord);
+  weight = max(1 - abs(xcoord - point), 0.0f);
+}
+
+__device__ bool between(int value, int lowerBound, int upperBound) {
+  return (value >= lowerBound && value <= upperBound);
+}
+
+__device__ void sumReduceShMem(volatile double s[]) {
+  /* obviously only works for 32 elements */
+  /* sums up a shared memory array of 32 elements, stores it in s[0] */
+  /* whole warp can then read first element (broadcasting) */
+  if (threadIdx.x < 16) { s[threadIdx.x] = s[threadIdx.x] + s[threadIdx.x+16]; }

```

Comment:
old gpus have 16 thread warps. See mshadow/gpu/reduce
you need to either fail on old gpus or add a sync here when running on old gpus